### PR TITLE
Refactoring: emptying handler's queues and using singleton instead of list

### PR DESCRIPTION
**Pull request by StrikingThirteen**
## Overview

Description of refactoring:
This is a refactoring on Server and Client HandshakeHandlers and MultipleReadRequest for the following purposes:
- Removing messages from queue at the time of writing and flushing them by the HandshakeHandlers
- Fixing unchecked assignment
- Replacing implementation of Queue (from ArrayDeque to LinkedList) for memory allocation and consistent o(1) 
- Using singleton instead of creating a list of one

Why should this be merged: 
- Avoiding to unnecessarily keep messages in queue of HandshakeHandler
- Consistent o(1) add/remove from queue and avoiding ArrayDeque's memory overhead
- Replacing unchecked assignment
- Avoiding list creation for one member

Related issue(s) (if applicable): #1316 

## Checklist (Definition of Done): 

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

**Comment by StrikingThirteen** [comment]
@annym Thanks Anny. Yes, it was green but I'll make sure that it passes again. The last I checked it failed on intermittent KillAndHeal.

**Comment by StrikingThirteen** [comment]
Travis tests are passing. Rebasing and squashing commits to prepare for merge.

**Comment by Maithem** [comment]
@StrikingThirteen  performance tests still running ?

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java [hunk 1]
```diff
@@ -30,7 +30,7 @@
     private final UUID nodeId;
     private final int handshakeTimeout;
     private final HandshakeState handshakeState;
-    private final Queue<CorfuMsg> messages = new ArrayDeque();
+    private final Queue<CorfuMsg> messages = new LinkedList<>();
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java]: I think I'm missing the point, but why is this change needed? A linked list implementation would be better if we remove elements that are not at the tail/head, but I don't see any logic that does that.

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java [hunk 2]
```diff
@@ -130,8 +129,8 @@ public void channelRead(ChannelHandlerContext ctx, Object m)
         log.info("channelRead: Handshake succeeded. Server Corfu Version: [{}]", corfuVersion);
         log.debug("channelRead: There are [{}] messages in queue to be flushed.", this.messages.size());
         // Flush messages in queue
-        for (CorfuMsg message : this.messages) {
-            ctx.writeAndFlush(message);
+        while (!messages.isEmpty()) {
+            ctx.writeAndFlush(messages.poll());
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java]: The messages are just being iterated, don't you actually want to drain the queue?

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java (hunk 1 above)
Comment (StrikingThirteen) [reply runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java]: LinkedList does provide a consistent time (without the amortization of ArrayDeque or without any additional backing array assigned). Also the messages queue is not being randomly accessed, it's purely a FIFO making LinkedList a natural pick. That being said, it's not required per se so if you're against it, I'll keep the ArrayDeque.

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java (hunk 2 above)
Comment (StrikingThirteen) [reply runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java]: Actually the queue is being drained and hence the poll(). Can you elaborate on why this is just an iteration?

Comment (Maithem) [reply runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java]: Hmm, this was suppose to be a comment on the diff:

for (CorfuMsg message : this.messages) {
            ctx.writeAndFlush(message);
}

Is this leaking actually a leak on master?

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java (hunk 1 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java]: I was referring to better caching behavior for arrays, but it seems like you introduced a remove (i.e. poll) which didn't exist in the old code. Also, this doesn't seem to be a critical path, so I'm in different.

Comment (StrikingThirteen) [reply runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java]: Sounds good. Thanks

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java (hunk 2 above)
Comment (StrikingThirteen) [reply runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java]: I see. correct, master still is not draining.
